File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.cc
#include <iostream>
#define fastio ios::sync_with_stdio(0);cin.tie(NULL);cout.tie(NULL);
#define MAX 1000001
using namespace std;
typedef long long LL;

LL arr[MAX];
LL tree[MAX*4];

LL build(int s, int e, int i) {
    if (s == e) return tree[i] = arr[s];

    int m = (s + e) / 2;
    return tree[i] = build(s, m, i * 2) + build(m + 1, e, i * 2 + 1);
}

LL interval_sum(int s, int e, int i, int l, int r) {
    if (e < l || r < s) return (LL)0;
    if (l <= s && e <= r) return tree[i];

    int m = (s + e) / 2;
    return interval_sum(s, m, i * 2, l, r) + interval_sum(m + 1, e, i * 2 + 1, l, r);
}

LL update(int s, int e, int i, int k, LL v) {
    if (k < s || e < k) return tree[i];
    if (s == e) return tree[i] = v;

    int m = (s + e) / 2;
    return tree[i] = update(s, m, i * 2, k, v) + update(m + 1, e, i * 2 + 1, k, v);
}

int main() {
    fastio
    int n, m, k; cin >> n >> m >> k;

    for (int i = 1; i <= n; i++) cin >> arr[i];
    build(1, n, 1);

    for (int i = 0; i < m + k; i++) {
        LL a, b, c; cin >> a >> b >> c;
        if (a == 1) update(1, n, 1, b, c);
        else cout << interval_sum(1, n, 1, b, c) << '\n';
    }
    return 0;
}

File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.py
# 2042 : 구간 합 구하기
import sys
input = sys.stdin.readline

n, m, k = map(int, input().rstrip().split())
arr = [0] + [int(input().rstrip()) for _ in range(n)]
tree = [0 for _ in range(n*4)]

def build(start, end, idx):
    if start == end:
        tree[idx] = arr[start]
        return
    
    mid = (start + end) // 2
    build(start, mid, idx * 2)
    build(mid + 1, end, idx * 2 + 1)
    tree[idx] = tree[idx * 2] + tree[idx * 2 + 1]

build(1, n, 1)

def interval_sum(start, end, idx, left, right):
    if end < left or right < start: return 0
    if left <= start and end <= right: return tree[idx]

    mid = (start + end) // 2
    return interval_sum(start, mid, idx * 2, left, right) + interval_sum(mid + 1, end, idx * 2 + 1, left, right)

def update(start, end, idx, k, x):
    if k < start or end < k: return
    if start == end:
        tree[idx] = x
        return
    
    mid = (start + end) // 2
    update(start, mid, idx * 2, k, x)
    update(mid + 1, end, idx * 2 + 1, k, x)
    tree[idx] = tree[idx * 2] + tree[idx * 2 + 1]

for _ in range(m+k):
    a, b, c = map(int, input().rstrip().split())
    if a == 1:
        update(1, n, 1, b, c)
    else:
        print(interval_sum(1, n, 1, b, c))